source/extensions/filters/network/echo/echo.h

```diff
@@ -5,12 +5,14 @@
 #include "common/common/logger.h"
 
 namespace Envoy {
-namespace Filter {
+namespace Extensions {
+namespace NetworkFilter {
+namespace Echo {
```

Comment:
What's the advantage of having such a deep namespace?  Do you find that more readable than the previous state?  Of course anything can be made to work, but I think without context it was harder for me to grok that 'Instance' resolved in Echo rather than in NetworkFilter or Extensions.

Reply:
This is actually flatter than it could be if it was [Envoy, Extensions, Filters, Network, Echo] as per the path.

Reply:
The reason I did what I did vs. the directory layout is that `Network` shadows the primary `Network` namespace and then requires a bunch of things to be prefixed with `Envoy::`. I thought this was better from a namespace perspective. Let me know if folks prefer the directory layout version and I will fix it.

re: namespace per filter, I feel we should definitely do this. Filters can get quite complicated and we may end up having a lot of them. I think they should be fully namespaced. `Instance` is done elsewhere, but I don't care if that is not desirable. I can rename to `EchoFilter` or something.

Reply:
@htuch I think in an earlier version of the Google C++ style guide there was a namespace-level per directory-level, but it has changed and I think the new preferred style is somewhat shallower. I think one of the concerns is confusion around names existing at multiple levels.

@mattklein123 in the past, with a similar challenge, we've relied mainly on the filter class for providing name-segregation.  Is there a common need for names belonging to a filter that are not in the class itself?  If so, then a namespace per filter is a reasonable solution.

This is mostly subjective though.  Big picture: LGTM.

Reply:
I really dislike "Instance" naming as IMO it makes it harder to find things (grep++) so if you're up for renaming and setting a precedence away from Instance, that'd be great.

I'd prefer as well to kill off class level namespacing in favor of descriptive class names - if we were willing to do FooFilter::Instance BarFilter::Instance I don't see that just having class FooFilter and class BarFilter is any more likely to cause conflicts and I prefer shorter namespaces.  I'm content to be overruled on this one but calling it out in case anyone other than Josh and I feel this way :-)

Reply:
I will get rid of `Instance` as I move things over, however I still feel pretty strongly that each filter/extension should be in a dedicated namespace. Some extensions will be quite complex (and already are). It seems silly to have to worry about multiple class conflicts in that case.

So in this case I would propose: `Envoy::Extensions::NetworkFilter::Echo::EchoFilter`.

I realize that is super contrived for such a simple case, but I think we should be consistent across all extensions and this is setting the stage for that.

Also, as long as we are in bikeshedding world, do we want the namespaces to be plural or not? `Extensions` vs. `Extension`?

Thoughts?

Reply:
Cool, as said I'm happy to lose the many-namespaces argument if I get to decrease Instances, so I'll call this current version a fine compromise :-)

I'm neutral on Extension vs Extensions so I'll let someone else bikeshed there :-P

Reply:
Yeah I think my primary objection was making it systematically hard to grep for things, so changing Echo::Instance to Echo::EchoFilter solves that well enough, so LGTM.

I'm still curious what are the common examples where you need to have a names associated with a filter outside the scope of the filter class.  One pattern I might expect is helper classes, but at least some of the time I'd expect them to be put in a common library, if they are potentially useful outside the context of the filter class.  An example of this might be the gzip library wrapper for the gzip filter, which might be useful outside the filter context, and would be great as a general library.  And otherwise as a nested class of the Filter class.

I think in the case of extensions it should be plural because there's expected to be more than one, but I don't feel too strongly.

Reply:
re: plurality, then `NetworkFilters` also?

re: namespace vs. nested class stuff. I realize at this point we are splitting hairs, but we already have examples of helper classes that are independently tested. E.g., the HTTP connection manager filter has many classes, as does the Mongo filter, Lua filter, etc. It just seems odd to box writers into potentially having conflicts or having to make everything a nested class. @envoyproxy/maintainers does anyone else have an opinion on this?

Reply:
Oh man, now that you've mentioned it I won't be able to unsee Extension*s* vs NetworkFilter.  Consistency between the two would be an improvement!

Reply:
I hadn't noticed.  I vote for NetworkFilters for the same reason as I voted for Extensions, but again don't feel too strongly.

Overall approach and desire for namespaces understood.  lgtm.